### DISCUSSION TITLE
Feature Request: Custom Line Series Labels

### DISCUSSION THREAD

**Discussion by christianrowlands** [patrykandpatrick/vico#550] [Ideas]
Thank you again for the great library. I released a version of my app a couple weeks ago using your 2.0.0-alpha.5 and it is working great. The app is open source, and here is a link to the source code in case you are interested to see how I am using your chart library: https://github.com/christianrowlands/android-network-survey

Anyway, I am working on adding another chart to the app listed above. In doing so, I came across the need to label each series directly on the chart, as well as in the legend. I of course tried setting the `dataLabel` for the LineSpec, but that just uses the y value as the label. I am hoping to provide a custom string for the series that is used to label the series.

Here is a screenshot of the chart as it stands right now:

<img width="466" alt="image" src="https://github.com/patrykandpatrick/vico/assets/8020686/6fcad876-979b-472b-8e33-b49f4800ba3c">

As you can see, I am using your library to visualize the usage of the Wi-Fi spectrum. Each line series represents a single Wi-Fi channel/network. Instead of displaying the 3 signal values as the data label, I am hoping to display the SSID (string) as the label. 

I found the code where you format that label [here](https://github.com/patrykandpatrick/vico/blob/e5d02a1e35e51c6da7e1d20d3359d9ae4bf7836f/vico/core/src/main/java/com/patrykandpatrick/vico/core/chart/layer/LineCartesianLayer.kt#L385). It uses the `dataLabelValueFormatter` to render the label text. I noticed that I could get the model from the chartValues, so I tried to extend `LineCartesianLayerModel` to see if I could store this information in a custom instance of the model, but I was unable to since it is not open. And even if I could, I would have a hard time mapping the SSID to the the proper series in `List<List<Entry>> series`.

```Kotlin
lineSpec.dataLabelValueFormatter.formatValue(
    value = chartEntry.y,
    chartValues = chartValues,
    verticalAxisPosition = verticalAxisPosition)
```

Is there a way that each series in the series List could have extra information stored? Maybe the series could be an object instead of just a `List<Entry>`? Then, the `lineSpec.dataLabelValueFormatter.formatValue` call could pass in another parameter of the `series`, and then the implementation can get whatever information they would need from the `Series` object, especially if they extended it to store whatever fields they want.

**Comment by christianrowlands** [comment]
As a reference, I am looking to create something very similar to what the Wi-Fi Analyzer app displays. Here is a screenshot of that app:

<img width="731" alt="image" src="https://github.com/patrykandpatrick/vico/assets/8020686/54ed4fc6-51f8-4144-b8ae-e08c30562059">

**Comment by Gowsky** [comment]
Hello @christianrowlands, we'll consider adding a built-in solution for such a feature, when there's more interest in it.
In the meantime, you can use existing APIs to add it yourself. Here's a summary of what needs to be done:

1. Custom labels (WiFi names) need to be put in some form (e.g. `List<String>`) to the  [`MutableExtraStore`](https://patrykandpatrick.com/vico/api/2.0.0-alpha.6/core/com.patrykandpatrick.vico.core.model/-mutable-extra-store/?query=class%20MutableExtraStore%20:%20ExtraStore). You can do that while updating series using [`Transaction#updateExtras`](https://patrykandpatrick.com/vico/api/2.0.0-alpha.6/core/com.patrykandpatrick.vico.core.model/-cartesian-chart-model-producer/-transaction/). Here's an example:
```kotlin
cartesianChartModelProducer.tryRunTransaction {
    updateExtras { store ->
        store[/*labels key*/] = /* list of labels */
     }
     lineSeries { 
         series(/* series 1 */) 
         series(/* series 2 */) 
         ...
     }
}
```
Unless labels are always placed in the middle x-axis value, you can also consider storing their corresponding x-axis and y-axis values.

2. A custom [`Decoration`](https://patrykandpatrick.com/vico/api/2.0.0-alpha.6/core/com.patrykandpatrick.vico.core.chart.decoration/-decoration/) could be used to draw the labels. [`onDrawAboveChart`](https://patrykandpatrick.com/vico/api/2.0.0-alpha.6/core/com.patrykandpatrick.vico.core.chart.decoration/-decoration/#1884458495%2FFunctions%2F186187275) yields [`ChartDrawContext`](https://patrykandpatrick.com/vico/api/2.0.0-alpha.6/core/com.patrykandpatrick.vico.core.chart.draw/-chart-draw-context/) which has access to [`ChartValues`](https://patrykandpatrick.com/vico/api/2.0.0-alpha.6/core/com.patrykandpatrick.vico.core.chart.values/-chart-values/?query=interface%20ChartValues), which has access to a [`CartesianChartModel`](https://patrykandpatrick.com/vico/api/2.0.0-alpha.6/core/com.patrykandpatrick.vico.core.model/-cartesian-chart-model/). The `ExtraStore` with custom labels is accessible from the `CartesianChartModel`. You can use [`ChartDrawContext#chartBounds`](https://patrykandpatrick.com/vico/api/2.0.0-alpha.6/core/com.patrykandpatrick.vico.core.chart.draw/-chart-draw-context/#-840910874%2FProperties%2F186187275), `ChartValues`, and `CartesianChartModel` to calculate coordinates for the labels.
3. `TextComponent` can be used to draw labels to a `Canvas`.
4. The custom `Decoration` needs to be added to a Chart. For Jetpack Compose use [`rememeberCartesianChart`](https://patrykandpatrick.com/vico/api/2.0.0-alpha.6/compose/com.patrykandpatrick.vico.compose.chart/remember-cartesian-chart?query=fun%20rememberCartesianChart(vararg%20layers:%20CartesianLayer%3C*%3E,%20startAxis:%20AxisRenderer%3CAxisPosition.Vertical.Start%3E?%20=%20null,%20topAxis:%20AxisRenderer%3CAxisPosition.Horizontal.Top%3E?%20=%20null,%20endAxis:%20AxisRenderer%3CAxisPosition.Vertical.End%3E?%20=%20null,%20bottomAxis:%20AxisRenderer%3CAxisPosition.Horizontal.Bottom%3E?%20=%20null,%20legend:%20Legend?%20=%20null,%20fadingEdges:%20FadingEdges?%20=%20null,%20decorations:%20List%3CDecoration%3E?%20=%20null,%20persistentMarkers:%20Map%3CFloat,%20Marker%3E?%20=%20null):%20CartesianChart) to add your decoration. For the view system use `CartesianChartView#chart#addDecoration`.

**Reply by christianrowlands** [reply]
I appreciate the approach to accomplish drawing the labels. I will give this a try and see how it goes.

Thank you!

**Reply by christianrowlands** [reply]
That worked. I ended up not using the extras store (although after I got done coding it I realized I should have used it, but oh well).

And here is a link to the source code just in case anyone reading this wants it as an example: https://github.com/christianrowlands/android-network-survey/blob/b98cb891e506d4d493385589412593a4d40ef144/networksurvey/src/main/java/com/craxiom/networksurvey/ui/wifi/SsidLabel.kt#L12

**Reply by Gowsky** [reply]
I'm glad it worked!
